Read /etc/login.defs

keszybz opened the pull request:
The bulk of the change here is a simplification and fixing of tests, and various adjustments to meson config.
The actual changes in the code are rather small.

There are two scenarios where we consider the system/user id boundary: when deciding whether to treat a user as "human", and when creating new users. The first scenario is quite simple: we simply need to look at /etc/login.defs and follow the config there. If /etc/login.defs is missing or doesn't specify that particular value, we use our built-in default. In the normal case our built-in default (which was created based on the /etc/login.defs installed on the system at build time) is completely equivalent to /etc/login.defs as provided by the distro. But if /etc/login.defs specifies some value, our builtin default must be ignored completely. The second scenario is more complicated, but it is only limited to sysusers, as this is the only place where we create users. (We synthesize root and nobody, but those have fixed numbers unrelated to /etc/login.defs.) In sysusers, we create users and groups with the same ids, so we use the same range for both, so we need some more complicated logic. See the commit that touches sysusers for details. The patch tries to honour /etc/login.defs, but also use values compatible with the compile-time defaults. This gives a mechanism for transition for the custom /etc/login.defs values to the distro defaults over time.

We also discussed using env vars to override the compile time defaults, instead of the tools parsing /etc/login.defs directly. I rejected this approach because there are many tools which would need to have the envvar set. Realistically, it would have to be set system-wide. But global envvars are shitty. It seems better to read the canonical file for this information. Also, something would have to set those variables, so we'd need to have some code to do that anyway, raising the overall complexity of the system. Finally, making sure that the env vars are set in places like package installation scriptlets is complicated, and again, the solution with the tools that need that reading /etc/login.defs is easier and more robust.

poettering commented:
For so many years we survived without without too much trouble. Why add so muc legacy support for making stuff configurable that shouldnt be configurable?

keszybz commented:
> For so many years we survived without without too much trouble. Why add so muc legacy support for making stuff configurable that shouldnt be configurable?

This isn't about configurability per se, but about transitions on existing systems and mixing builds from different vendors. I wrote up a summary in the commit 3/7. In addition to what is written there, I think it's fair to say that this is a problem that comes with popularity of systemd and deployment on large-scale systems. On my laptop, even if I had to renumber users, I would be able to do this without trouble. But on a large-scale deployment, problems arise. Also, multiple reports opened about this show that this is a problem, e.g. #3855 is from 2016.

keszybz commented:
Hmm, bionic-arm64 is showing test results for the previous version of the patch... I'll force-push again.

keszybz commented:
I updated the patch that converts TEST-21-SYSUSERS to a test-sysusers.sh to install the test when install-tests=true. Let's see if that helps with the CI.

keszybz commented:
That's is strange. The test is run from /var/tmp or /tmp. *Some* of the files are present (because the glob test passes), but not others. The best guess I have is that something is deleting those files or that the transfer is incomplete. The meson installation mechanism copies over the whole directory, so I don't see how it could miss some files but not others.

poettering commented on src/shared/user-record.c:
```diff
@@ -48,11 +50,29 @@ static int read_login_defs(login_defs *ret_defs, const char *path) {
+                        r = parse_uid(t, &sys_uid_min);
+                        if (r < 0)
+                                log_debug_errno(r, "%s: failed to parse SYS_UID_MIN, ignoring: %m", path);
+                        if (sys_uid_min == 0) {
```
hmm, i don#t think we should correct 0 → 101 here. This is about allocation, and I think in most setups allocating from the whole < 1000 range is totally ok. i.e. it might make sense to bump 0 to 1 here, since we know for sure that uid 0 is taken by some user and hence not useful as space to allocate from, but conceptually from the distance I think it totally makes sense to say 0…999 is the system uid range that can be allocated from, even though we know that 0 is definitely always already taken...

Hence, maybe upgrade sys_uid_min=0 → sys_uid_min=1, as kind of an optimization for the allocation algorithms, but beyond that i wouldn't do anything here.

poettering commented on src/sysusers/sysusers.c:
```diff
@@ -1949,10 +1950,26 @@ static int run(int argc, char *argv[]) {
+                const char *path;
+
+                path = prefix_roota(arg_root, "/etc/login.defs");
+                r = read_login_defs(&defs, path);
```
hmm, we have been a bit sloppy with insisting on chase_symlinks() for opening the files in sysusers.c relative to the --root= dir, but I think we should become better at that and at least for new code always use chase_symlinks() so that we don't get confused by /etc/ in the image being a symlink.

In fact we should probably add chase_symlinks_and_fopen(), similar to chase_symlinks_and_open() and use that everywhere. Not doing that for the password databases is actually quite problematic security wise, given that this means invoking the tool on rogue images will possibly manipulate the host password database...

I know this problem is unrelated to the patch at hand, but this has been bugging me for a while, ad for new code I think we should insist on this

poettering commented on src/sysusers/sysusers.c:
```diff
@@ -1949,10 +1950,26 @@ static int run(int argc, char *argv[]) {
+                 * /etc/login.defs. That way the uids/gids which we allocate will be interpreted correctly,
+                 * even if /etc/login.defs is removed later. (The bottom bound doesn't matter much, since
+                 * it's only used during allocation, so we use the configured value directly). */
+                uid_t beg = defs.sys_uid_min,
```
why abbreviate begin to beg? those two extra chars are not costly...

keszybz commented on src/shared/user-record.c:
```diff
@@ -21,6 +23,85 @@
+        if (r == -ENOENT)
+                goto assign;
+        if (r < 0)
+                return log_debug_errno(r, "Failed to open %s: %m", path);
```
It'll only fail when called from tests or from sysusers (when added there in one of the later patches). If we can't read the config file then something is seriously wrong. I'm pretty sure we need to treat this is an error when creating new users. If someone added the config file, they probably need it and rely on the config there to be honoured.

keszybz replied:
I reworked this to never fail. (sysusers users a slightly different code path.)

keszybz commented on src/shared/user-record.h:
```diff
@@ -41,6 +36,14 @@ static inline bool gid_is_container(gid_t gid) {
+typedef struct login_defs {
+        uid_t sys_uid_max;
+        gid_t sys_gid_max;
+        bool _initialized;
```
I moved the boolean out of the structure and into a static var in the function. I think that's nicer.

keszybz commented on src/shared/user-record.c:
```diff
@@ -21,6 +23,85 @@
+        return 0;
+}
+
+int acquire_login_defs(const login_defs **ret_defs, bool ignore_failure) {
```
I think we should ignore failure when querying the value (e.g. from journald when deciding whether to split a journal). But from sysusers, I think we should fail. After all it's called for an admin action and if the config files are inaccessible, it's something to fix not ignore.

keszybz commented on meson.build:
```diff
@@ -3538,12 +3528,10 @@ status = [
-        'maximum dynamic UID:               @0@'.format(dynamic_uid_max),
-        'minimum container UID base:        @0@'.format(container_uid_base_min),
-        'maximum container UID base:        @0@'.format(container_uid_base_max),
+        'system UIDs:                       @0@–@1@'.format(conf.get('SYSTEM_UID_MIN'), conf.get('SYSTEM_UID_MAX')),
```
I thought we only use .. because it's hard to type n-dash on a keyboard?

poettering replied:
I use … mostly because it's so pretty. (and kinda how mathematicians write ranges, no?)

(i mean, ignore me on this, it really doesn't matter at all...)

keszybz commented on meson.build:
```diff
@@ -694,35 +694,25 @@ if time_epoch == -1
-                system_gid_max = 999
-        else
-                system_gid_max = system_gid_max.to_int()
+foreach tuple : [['system-uid-min', 'SYS_UID_MIN', 101],
```
login.defs(5) says so. The way I understand this is that the range *below* SYS_UID_MIN is reserved for statically allocated UIDs. We do this with a few gids defined in `sysusers.d/basic.conf`. Fedora has a list of ~190 such uids defined in setup.rpm, so it defines SYS_UID_MIN=201 to not overlap.

poettering replied:
this sounds like historic stuff mostly. I doubt we should bother on legacyfree systems with this. i.e. let's not carry this on needlessly. Let#s set this to 1 here, I see little reason to keep a separate range for thta stuff...

keszybz commented:
I reworked this quite a bit. I hope I didn't miss anything:
- meson options are now `-Dsystem-{uid,gid}-max` (as before) and `-Dsystem-alloc-{uid,gid}-min`
- UIDS-GIDS.txt makes it clear that modifying login.defs is not recommended
- the fields in the structure are renamed to `systemd_{uid,gid}_max`, `systemd_alloc_{uid,gid}_min`
- I added chase_symlinks_and_fopen_unlocked(), and sysusers now uses it
- there is more tests
- ignore_errors in acquire_login_defs is gone
- when not otherwise specified, the default allocation range is 1–999.
- in meson output, the format is `<=998 (alloc >=201)` to not confused people that uids below the lower bound are not system uids
- I also make some small cleanups in other tests
- `._initialized` field is gone, replaced by a local variable in the function

keszybz commented:
Oh, I didn't add any printed warning, because I don't see a particularly good place to put it. Doing it at every boot would be annoying. I think the label in the docs should be enough. After all, most people didn't feel the urge to modify login.defs wantonly, so I don't think they'll start now.

poettering commented on src/shared/user-record.c:
```diff
@@ -21,6 +23,98 @@
+#endif
+
+int acquire_login_defs(const login_defs **ret_defs) {
+        static thread_local login_defs defs = {
```
```c
#if ENABLE_COMPAT_MUTABLE_UID_BOUNDARIES
static thead_local login_defs defs = {
#else
static const login_defs defs = {
#endif
…
```

poettering commented on src/shared/user-record.c:
```diff
@@ -21,6 +23,98 @@
+        if (r < 0)
+                return log_debug_errno(r, "%s: failed to parse %s %s, ignoring: %m", path, name, t);
+        else if (uid == 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "%s: %s == 0, ignoring.", path, name);
```
i still think uid == 0 is actually fine, and should not result in any log message. it might be OK to upgrade it to 1 implicitly, simply to optimize the alloc argument a bit, but as mentioned i think saying 0…999 is the system user range is totally OK, hence 0 shouldn't be refused here, even if we know that 0 is going to be root anyway... (also see earlier comment on this)

poettering commented on src/shared/user-record.c:
```diff
@@ -21,6 +23,98 @@
+static int read_login_defs(login_defs *ret_defs, const char *path) {
+        _cleanup_fclose_ FILE *f = NULL;
+        uid_t sys_uid_max = SYSTEM_UID_MAX;
+        gid_t sys_gid_max = SYSTEM_GID_MAX;
```
maybe make these two an instance of `struct login_defs` here too? no need to allocate these fields separately on the stack, if you can just allocate the whole structure in one, after all you need all its fields here anyway. Also makes the exit path simpler, as you can just copy the whole struct into teh return buffer.

poettering commented on src/shared/user-record.h:
```diff
@@ -41,6 +36,13 @@ static inline bool gid_is_container(gid_t gid) {
+typedef struct login_defs {
+        uid_t system_uid_max;
+        gid_t system_gid_max;
+} login_defs;
```
btw, i think we kinda followed the rules that typedefs either use CamelCase naming, or carry the "sd_" prefix. but not sure, maybe we don't follow that rule strictly, but i think this would be an outlier here...

poettering commented on src/shared/user-record.c:
```diff
@@ -21,6 +23,98 @@
+         * probably not for sysusers. */
+
+#if ENABLE_COMPAT_MUTABLE_UID_BOUNDARIES
+        static thread_local int initialized;
```
couldn#t find the document that states thread local vars are zero-initialized implicitly. Are you sure they are?

keszybz replied:
I added initalization now.

poettering commented on src/shared/user-record.c:
```diff
@@ -21,6 +23,98 @@
+}
+#endif
+
+int acquire_login_defs(const login_defs **ret_defs) {
```
since the function cannot fail, maybe just return the structure pointer directly, instead of using return-by-param?

poettering commented on src/shared/user-record.c:
```diff
@@ -37,19 +37,22 @@ static int parse_alloc_uid(const char *path, const char *name, const char *t, ui
 
-static int read_login_defs(login_defs *ret_defs, const char *path) {
-        _cleanup_fclose_ FILE *f = NULL;
+int read_login_defs(login_defs *ret_defs, const char *path, const char *root) {
```
the "path" argument is always the same no? maybe drop it, and just take the root?

keszybz replied:
It's used in tests...

poettering commented:
> Oh, I didn't add any printed warning, because I don't see a particularly good place to put it. Doing it at every boot would be annoying. I think the label in the docs should be enough. After all, most people didn't feel the urge to modify login.defs wantonly, so I don't think they'll start now.

I think doing it in systemd-sysusers would be best, on the first allocation. This is were we actually make use of this for allocation purposes, and it also means you will only see this when new system users are added, i.e. at worst after a package installation. It would not be shown on every boot even.

keszybz commented:
Updated with the requested changes. The last commit prints a warning, PTAL.

poettering commented on src/sysusers/sysusers.c:
```diff
@@ -84,6 +84,9 @@ static uid_t search_uid = UID_INVALID;
 static unsigned n_uid_range = 0;
 
+static UGIDAllocationRange login_defs = {};
+static bool login_defs_need_warning = false;
```
i wished we'd lose global vars, no add more... we should wrap the whole shebang here in a "Context" or "Manager" structure one of those days...

keszybz commented:
bionic-i386:
```
autopkgtest [22:50:12]: ERROR: timed out on command "su -s /bin/bash root -c set -e; export USER=`id -nu`; . /etc/profile >/dev/null 2>&1 || true;  . ~/.profile >/dev/null 2>&1 || true; buildtree="/tmp/autopkgtest.BVqnjp/build.2AP/systemd"; mkdir -p -m 1777 -- "/tmp/autopkgtest.BVqnjp/upstream-artifacts"; export AUTOPKGTEST_ARTIFACTS="/tmp/autopkgtest.BVqnjp/upstream-artifacts"; export ADT_ARTIFACTS="$AUTOPKGTEST_ARTIFACTS"; mkdir -p -m 755 "/tmp/autopkgtest.BVqnjp/autopkgtest_tmp"; export AUTOPKGTEST_TMP="/tmp/autopkgtest.BVqnjp/autopkgtest_tmp"; export ADTTMP="$AUTOPKGTEST_TMP"; export DEBIAN_FRONTEND=noninteractive; export LANG=C.UTF-8; export DEB_BUILD_OPTIONS=parallel=1; unset LANGUAGE LC_CTYPE LC_NUMERIC LC_TIME LC_COLLATE   LC_MONETARY LC_MESSAGES LC_PAPER LC_NAME LC_ADDRESS   LC_TELEPHONE LC_MEASUREMENT LC_IDENTIFICATION LC_ALL;rm -f /tmp/autopkgtest_script_pid; set -C; echo $$ > /tmp/autopkgtest_script_pid; set +C; trap "rm -f /tmp/autopkgtest_script_pid" EXIT INT QUIT PIPE; cd "$buildtree"; export AUTOPKGTEST_NORMAL_USER=ubuntu; export ADT_NORMAL_USER=ubuntu; export 'CFLAGS=-O0'; export 'DEB_BUILD_PROFILES=noudeb'; export 'TEST_UPSTREAM=1'; export 'CONFFLAGS_UPSTREAM=--werror -Dslow-tests=true'; export 'UPSTREAM_PULL_REQUEST=17172'; export 'GITHUB_STATUSES_URL=https://api.github.com/repos/systemd/systemd/statuses/4b6f9b202e0d2edaa32a2520534be33dcba3be39'; chmod +x /tmp/autopkgtest.BVqnjp/build.2AP/systemd/debian/tests/upstream; touch /tmp/autopkgtest.BVqnjp/upstream-stdout /tmp/autopkgtest.BVqnjp/upstream-stderr; /tmp/autopkgtest.BVqnjp/build.2AP/systemd/debian/tests/upstream 2> >(tee -a /tmp/autopkgtest.BVqnjp/upstream-stderr >&2) > >(tee -a /tmp/autopkgtest.BVqnjp/upstream-stdout);" (kind: test)
```
Looks like the usual timeout.
